### PR TITLE
feat: Leverage charts

### PR DESCRIPTION
**Pull request by 0xonramp**
Uses visx for leverage charts.

### CODE REVIEW

File: src/app/leverage/page.tsx [hunk 1]
```diff
@@ -31,7 +34,17 @@ export default function Page() {
...
+  const { baseToken, baseTokens, indexToken, onSelectBaseToken } =
+    useLeverageToken()
+  const [currentTab, setCurrentTab] = useState<ChartTab>('indexcoop-chart')
+
+  useEffect(() => {
+    document.body.classList.add('bg-ic-dark')
+    return () => {
+      document.body.classList.remove('bg-ic-dark')
+    }
+  })
```
Comment (0xonramp) [inline src/app/leverage/page.tsx:46]: Design request to ensure the background is dark when user scrolls/pulls down on mobile, so background needs to be applied to `body`. The root layout is 'light', so we would have needed to create two separate route groups for dark and light, and split the code by route group.

Not ideal, but I think this is preferred

File: src/lib/hooks/use-query-params.ts [hunk 2]
```diff
@@ -99,7 +99,7 @@ export const useQueryParams = <T extends Partial<UseQueryParamsArgs>>(
         network?.toString() ?? queryParams.get('network') ?? '',
       )
 
-      router.replace(`?${queryParams.toString()}`)
+      router.replace(`?${queryParams.toString()}`, { scroll: false })
```
Comment (0xonramp) [inline src/lib/hooks/use-query-params.ts:102]: Avoid scrolling when user changes chain/token.

File: src/app/leverage/components/chart-tabs.tsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+  return (
+    <nav
+      className='divide-ic-blue-700 mt-2 flex divide-x rounded-t-lg'
+      aria-label='Tabs'
+    >
+      {tabs.map((tab, tabIdx) => (
+        <button
+          key={tab.name}
+          className={clsx(
```
Comment (tokdaniel) [inline src/app/leverage/components/chart-tabs.tsx:37]: note: you can use cn here, for a tailwind specific merging of classes too

File: src/app/leverage/provider.tsx [hunk 4]
```diff
@@ -177,6 +184,25 @@ export function LeverageProvider(props: { children: any }) {
     indexTokenAddresses,
   )
 
+  const {
+    data: { nav, navchange },
+  } = useQuery({
+    enabled: isAddress(indexToken.address ?? ''),
```
Comment (tokdaniel) [inline src/app/leverage/provider.tsx:190]: you can use tokenlist's isAddress, so you dont need the null coalesce

File: src/components/charts/price-chart.tsx [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import { formatDollarAmount } from '@/lib/utils'
+import { getAddressForToken } from '@/lib/utils/tokens'
+
+type Props = {
+  indexToken: Token
+}
+
+export function PriceChart({ indexToken }: Props) {
+  const { chainId } = useNetwork()
+  const tokenAddress = getAddressForToken(
```
Comment (tokdaniel) [inline src/components/charts/price-chart.tsx:17]: also, here you can utilize the new tokenlist's stuff

File: src/components/charts/use-price-chart-data.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+      const data = await fetchTokenHistoricalData({
+        tokenAddress: indexTokenAddress!,
+        ...fetchSettings,
+      })
+      const historicalData = data
+        ?.map((datum) => ({
+          ...datum,
+          NetAssetValue: Number(datum.NetAssetValue?.toFixed(2)),
+        }))
+        .sort(
```
Comment (tokdaniel) [inline src/components/charts/use-price-chart-data.ts:59]: inverted sorting here wouldn't remove the need for the reverse?

Comment (0xonramp) [reply src/components/charts/use-price-chart-data.ts:59]: good call - updating

File: src/app/leverage/components/your-tokens.tsx [hunk 7]
```diff
@@ -106,7 +106,7 @@ export function YourTokens() {
                   (selected || hover) && 'text-ic-white',
                 )}
               >
-                Open Positions
+                Your Tokens
```
Comment (0xonramp) [inline src/app/leverage/components/your-tokens.tsx:109]: @tokdaniel FYI - this was a request from Alica

File: src/app/leverage/components/stats.tsx [hunk 8]
```diff
@@ -65,18 +62,7 @@ function StatsItem(props: StatsItemProps) {
...
-                xmlns='http://www.w3.org/2000/svg'
-                width='8'
-                height='6'
-                viewBox='0 0 8 6'
-                fill='none'
-              >
-                <path d='M0.666667 5.86914L4 2.26083L7.33333 5.86914L8 5.14748L4 0.817505L6.24448e-08 5.14748L0.666667 5.86914Z' />
-              </svg>
-            </div>
+            <ChevronUpIcon className={clsx(iconColor, iconScale)} />
```
Comment (tokdaniel) [inline src/app/leverage/components/stats.tsx:65]: just fyi: we have another function now called `cn` which also makes sure, that the classes are merged in correct order, while using clsx too.

File: src/app/leverage/components/stats.tsx [hunk 9]
```diff
@@ -29,11 +26,11 @@ export function Stats() {
...
       <StatsItem
         token={indexToken}
-        change24h={''}
-        change24hIsPositive={true}
+        change24h={navchange !== 0 ? formatPercentage(navchange) : ''}
+        change24hIsPositive={navchange > 0}
         high24h={high24h}
         low24h={low24h}
-        price={indexTokenPrice}
+        price={nav > 0 ? formatDollarAmount(nav) : ''}
```
Comment (janndriessen) [inline src/app/leverage/components/stats.tsx:33]: minor: all these changes could be moved into the formatting hook which would make this look a lil cleaner

File: src/app/leverage/types.ts [hunk 10]
```diff
@@ -27,3 +27,5 @@ export interface LeverageToken extends Token {
   baseToken: string
   leverageType: LeverageType
 }
+
+export type ChartTab = 'indexcoop-chart' | 'tradingview-chart'
```
Comment (janndriessen) [inline src/app/leverage/types.ts:31]: minor: would prefer enum

**Review by janndriessen — APPROVED** [review]
🌟

File: src/app/leverage/components/stats.tsx (hunk 9 above)
Comment (0xonramp) [reply src/app/leverage/components/stats.tsx:33]: @janndriessen good call - i'll update this one and other usages in a follow-up PR

File: src/app/leverage/types.ts (hunk 10 above)
Comment (0xonramp) [reply src/app/leverage/types.ts:31]: let's discuss more - I know @tokdaniel has strong opinions here too

**Pull request merged**